### PR TITLE
Backend filtering support for the new publications page (root controller)

### PR DESCRIPTION
**Pull request by mtaylorgds**
Update the new root controller, which uses the GOV.UK Design System layout template, to accept various submitted filter parameters, which are then used to make a filtered list of editions available to the rendered view, via a presenter.

There is a very basic, unstyled UI, (using basic rails controls rather than Design System components) for the purposes of demonstrating that the filters work / manual testing. This UI is simply there for test purposes and will be replaced by the separate ongoing work that will create the filter component and the results table component.

The controller is behind a feature flag, and so none of this is visible to general users.

[Trello ticket](https://trello.com/c/wucvIcYg)

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

### CODE REVIEW

**Review by davidtrussler — APPROVED** [review]
Hi @mtaylorgds - this looks good to me from the perspective of providing the basis for the table. The code looks good but would be prefer someone with more Rails knowledge to also approve.

**Comment by syed-ali-tw** [comment]
very minor.
Slight mismatch in the commit message vs actual changes for the first commit.
"This change is only at the controller end" whereas the changes are backend changes beyond controller.

File: app/controllers/root_controller.rb [hunk 1]
```diff
@@ -3,7 +3,27 @@
...
+    ready
+    scheduled_for_publishing
+    published
+    archived
+  ].freeze
+
   def index
-    @presenter = FilteredEditionsPresenter.new
+    states_filter_params = filter_params.to_h[:states_filter]
+    sanitised_states_filter_params = states_filter_params&.select { |fp| PERMITTED_FILTER_STATES.include?(fp) }
```
Comment (syed-ali-tw) [inline app/controllers/root_controller.rb:21]: not sure if there is any sanitisation on states filter here, seems more like filtered states params

File: test/unit/presenters/filtered_editions_presenter_test.rb [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      published_guide = FactoryBot.create(:guide_edition, state: "published")
+
+      filtered_editions = FilteredEditionsPresenter.new.editions
+
+      assert_equal(2, filtered_editions.count)
+      assert_equal(draft_guide, filtered_editions[0])
+      assert_equal(published_guide, filtered_editions[1])
+    end
+
+    should "filter by specified states" do
```
Comment (syed-ali-tw) [inline test/unit/presenters/filtered_editions_presenter_test.rb:18]: could just be should "filter by 'state'"
like other tests

File: test/unit/presenters/filtered_editions_presenter_test.rb [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+      assert_equal([assigned_to_anna], filtered_editions)
+    end
+
+    should "filter by 'not assigned'" do
+      anna = FactoryBot.create(:user, name: "Anna")
+      FactoryBot.create(:guide_edition, assigned_to: anna.id)
+      not_assigned = FactoryBot.create(:guide_edition)
+
+      filtered_editions =
```
Comment (syed-ali-tw) [inline test/unit/presenters/filtered_editions_presenter_test.rb:44]: could be single line similar to line 34

File: test/unit/presenters/filtered_editions_presenter_test.rb [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      assert_equal([not_assigned], filtered_editions)
+    end
+
+    should "ignore invalid 'assigned to'" do
+      anna = FactoryBot.create(:user, name: "Anna")
+      FactoryBot.create(:guide_edition, assigned_to: anna.id)
+      FactoryBot.create(:guide_edition)
+
+      filtered_editions =
+        FilteredEditionsPresenter.new(assigned_to_filter: "not a valid user id").editions
```
Comment (syed-ali-tw) [inline test/unit/presenters/filtered_editions_presenter_test.rb:56]: could be single line similar to line 34

File: test/unit/presenters/filtered_editions_presenter_test.rb [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        FilteredEditionsPresenter.new(format_filter: "guide").editions
+
+      assert_equal([guide], filtered_editions)
+    end
+
+    should "return all formats when specified by the format filter" do
+      FactoryBot.create(:guide_edition)
+      FactoryBot.create(:completed_transaction_edition)
+
+      filtered_editions =
```
Comment (syed-ali-tw) [inline test/unit/presenters/filtered_editions_presenter_test.rb:75]: could be single line similar to line 34

File: test/unit/presenters/filtered_editions_presenter_test.rb [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        FilteredEditionsPresenter.new(format_filter: "all").editions
+
+      assert_equal(2, filtered_editions.count)
+    end
+
+    should "filter by a partially-matching title" do
+      guide_fawkes = FactoryBot.create(:guide_edition, title: "Guide Fawkes")
+      FactoryBot.create(:guide_edition, title: "Hitchhiker's Guide")
+
+      filtered_editions =
```
Comment (syed-ali-tw) [inline test/unit/presenters/filtered_editions_presenter_test.rb:85]: could be single line similar to line 34

**Comment by syed-ali-tw** [comment]
LGTM.
few very minor comments.

File: app/presenters/filtered_editions_presenter.rb [hunk 7]
```diff
@@ -1,7 +1,59 @@
...
+  end
+
+  def available_users
+    User.enabled.alphabetized
+  end
+
   def editions
-    Edition.all
+    result = editions_by_format
+    result = apply_states_filter(result)
```
Comment (syed-ali-tw) [inline app/presenters/filtered_editions_presenter.rb:17]: might be difficult to do anything about it, not sure if I like passing result and reading result.

**Comment by mtaylorgds** [comment]
> very minor. Slight mismatch in the commit message vs actual changes for the first commit. "This change is only at the controller end" whereas the changes are backend changes beyond controller.

Not really a mismatch—that's what I meant by the message. I've reworded to replace "controller-end" with "backend".

File: app/controllers/root_controller.rb (hunk 1 above)
Comment (mtaylorgds) [reply app/controllers/root_controller.rb:21]: The sanitisation is the removal of any states that aren't valid.

File: test/unit/presenters/filtered_editions_presenter_test.rb (hunk 4 above)
Comment (mtaylorgds) [reply test/unit/presenters/filtered_editions_presenter_test.rb:54]: This line would be too long, so I have left it as-is.

File: app/presenters/filtered_editions_presenter.rb (hunk 7 above)
Comment (mtaylorgds) [reply app/presenters/filtered_editions_presenter.rb:17]: Yeah, I get what you mean, but I'm not sure how else to achieve this really.

**Review by syed-ali-tw — APPROVED** [review]

**Pull request merged**